File: frontend/src/lib/types/dnd.ts
// Do Not Disturb (DND) Types

export interface DNDSettings {
  id?: string;
  user_id: string;
  enabled: boolean;
  schedule: DNDSchedule;
  overrides: DNDOverride[];
  allow_p1_override: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface DNDSchedule {
  weekly: DNDTimeSlot[];
  timezone: string;
}

export interface DNDTimeSlot {
  day: DayOfWeek;
  start: string; // HH:MM format (24-hour)
  end: string; // HH:MM format (24-hour)
}

export type DayOfWeek =
  | 'sunday'
  | 'monday'
  | 'tuesday'
  | 'wednesday'
  | 'thursday'
  | 'friday'
  | 'saturday';

export interface DNDOverride {
  start: string; // ISO date string
  end: string; // ISO date string
  reason?: string;
}

export interface UpdateDNDSettingsRequest {
  enabled?: boolean;
  schedule?: DNDSchedule;
  overrides?: DNDOverride[];
  allow_p1_override?: boolean;
}

export interface AddDNDOverrideRequest {
  start: string; // ISO date string
  end: string; // ISO date string
  reason?: string;
}

export interface DNDStatusResponse {
  in_dnd_mode: boolean;
  priority: string;
}

// Helper constants
export const DAYS_OF_WEEK: DayOfWeek[] = [
  'sunday',
  'monday',
  'tuesday',
  'wednesday',
  'thursday',
  'friday',
  'saturday',
];

export const DAY_LABELS: Record<DayOfWeek, string> = {
  sunday: 'Sunday',
  monday: 'Monday',
  tuesday: 'Tuesday',
  wednesday: 'Wednesday',
  thursday: 'Thursday',
  friday: 'Friday',
  saturday: 'Saturday',
};

// Common timezone options
export const COMMON_TIMEZONES = [
  'UTC',
  'America/New_York',
  'America/Chicago',
  'America/Denver',
  'America/Los_Angeles',
  'America/Toronto',
  'America/Vancouver',
  'Europe/London',
  'Europe/Paris',
  'Europe/Berlin',
  'Europe/Amsterdam',
  'Asia/Tokyo',
  'Asia/Shanghai',
  'Asia/Singapore',
  'Asia/Dubai',
  'Asia/Kolkata',
  'Australia/Sydney',
  'Australia/Melbourne',
  'Pacific/Auckland',
];

File: backend/internal/core/domain/dnd.go
package domain

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// UserDNDSettings represents a user's Do Not Disturb configuration
type UserDNDSettings struct {
	ID              uuid.UUID
	UserID          uuid.UUID
	Enabled         bool
	Schedule        json.RawMessage
	Overrides       json.RawMessage
	AllowP1Override bool
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

// DNDSchedule represents the weekly schedule for DND
type DNDSchedule struct {
	Weekly   []DNDTimeSlot `json:"weekly"`
	Timezone string        `json:"timezone"` // IANA timezone string
}

// DNDTimeSlot represents a single time slot in the DND schedule
type DNDTimeSlot struct {
	Day   string `json:"day"`   // monday, tuesday, wednesday, thursday, friday, saturday, sunday
	Start string `json:"start"` // HH:MM format (24-hour)
	End   string `json:"end"`   // HH:MM format (24-hour)
}

// DNDOverride represents a one-time DND override period
type DNDOverride struct {
	Start  time.Time `json:"start"`
	End    time.Time `json:"end"`
	Reason string    `json:"reason,omitempty"`
}

// ParseSchedule parses the raw JSON schedule into a structured format
func (s *UserDNDSettings) ParseSchedule() (*DNDSchedule, error) {
	if len(s.Schedule) == 0 {
		return &DNDSchedule{}, nil
	}
	var schedule DNDSchedule
	if err := json.Unmarshal(s.Schedule, &schedule); err != nil {
		return nil, err
	}
	return &schedule, nil
}

// ParseOverrides parses the raw JSON overrides into a structured format
func (s *UserDNDSettings) ParseOverrides() ([]DNDOverride, error) {
	if len(s.Overrides) == 0 {
		return []DNDOverride{}, nil
	}
	var overrides []DNDOverride
	if err := json.Unmarshal(s.Overrides, &overrides); err != nil {
		return nil, err
	}
	return overrides, nil
}

// Valid day names
var validDays = map[string]int{
	"sunday":    0,
	"monday":    1,
	"tuesday":   2,
	"wednesday": 3,
	"thursday":  4,
	"friday":    5,
	"saturday":  6,
}

// IsValidDay checks if a day name is valid
func IsValidDay(day string) bool {
	_, ok := validDays[day]
	return ok
}

// GetDayIndex returns the weekday index for a day name (0 = Sunday)
func GetDayIndex(day string) int {
	return validDays[day]
}
